File: js/src/program/ops/opInterpolatedString.js
import { call } from './utils';

export default {
  /** { interpolations: [{ before: string, pipe: [op] }], after: string } */
  async op(runtime, input, params, scope, next) {
    const interpolations = await runtime.muxAsync([params.interpolations], (interpolation) =>
      runtime.executeInstructions(interpolation.pipe, [input], scope, (output) => [
        interpolation.before + runtime.stringifyJSON(output, true),
      ]),
    );

    return runtime.muxAll(interpolations, (...parts) => next(parts.join('') + params.after, scope));
  },

  /** { interpolations: [{ before: string, value: function }], after: string } */
  map(runtime, params) {
    return {
      interpolations: runtime.muxOne([params.interpolations], (entry) => ({
        before: runtime.assertType(params.before, 'string'),
        pipe: call(entry.value),
      })),
      after: runtime.assertType(params.after, 'string'),
    };
  },
};

File: go/program/ops/opInterpolatedString.go
package ops

import (
	"github.com/2manyvcos/jpl/go/definition"
	"github.com/2manyvcos/jpl/go/jpl"
)

type opInterpolatedString struct{}

// { interpolations: [{ before: string, pipe: [op] }], after: string }
func (opInterpolatedString) OP(runtime jpl.JPLRuntime, input any, params definition.JPLInstructionParams, scope jpl.JPLRuntimeScope, next jpl.JPLScopedPiper) ([]any, jpl.JPLError) {
	panic("TODO:")
}

// { interpolations: [{ before: string, pipe: function }], after: string }
func (opInterpolatedString) Map(runtime jpl.JPLRuntime, params jpl.JPLInstructionParams) (definition.JPLInstructionParams, jpl.JPLError) {
	panic("TODO:")
}
